vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,948 @@
+    pub fn new(spi_ctrl: Arc<SpiController<D>>) -> Result<Self> {
+        Ok(VhostUserSpiBackend {
+            spi_ctrl,
+            event_idx: false,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            mem: None,
+        })
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_requests(
+        &self,
+        requests: Vec<SpiDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        let mut reqs: Vec<SpiTransReq> = Vec::new();
+
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        // Iterate over each SPI request and push it to "reqs" vector.
+        for desc_chain in requests.clone() {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if (descriptors.len() != 3) && (descriptors.len() != 4) {
+                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+            }
+
+            if descriptors.len() == 4
+                && (descriptors[1].is_write_only() || !descriptors[2].is_write_only())
+            {
+                return Err(Error::TxRxBufOrderMismatch);
+            }
+
+            let desc_out_hdr = descriptors[0];
+
+            if desc_out_hdr.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0));
+            }
+
+            if desc_out_hdr.len() as usize != size_of::<VirtioSpiTransferHead>() {
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<VirtioSpiTransferHead>(),
+                    desc_out_hdr.len(),
+                ));
+            }
+
+            let out_hdr = desc_chain
+                .memory()
+                .read_obj::<VirtioSpiTransferHead>(desc_out_hdr.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let desc_in_hdr = descriptors[descriptors.len() - 1];
+            if !desc_in_hdr.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(descriptors.len() - 1));
+            }
+
+            if desc_in_hdr.len() as usize != size_of::<u8>() {
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<u8>(),
+                    desc_in_hdr.len(),
+                ));
+            }
+
+            let mut tx_buf: Vec<u8>;
+            let mut rx_buf: Vec<u8>;
+            let trans_len: u32;
+
+            (tx_buf, rx_buf, trans_len) = match descriptors.len() {
+                3 => {
+                    // half-duplex transfer
+                    let desc_buf = descriptors[1];
+                    let len = desc_buf.len();
+
+                    if len == 0 {
+                        return Err(Error::UnexpectedDescriptorSize(1, len));
+                    }
+
+                    if desc_buf.is_write_only() {
+                        rx_buf = vec![0; len as usize];
+                        tx_buf = Vec::<u8>::new();
```

Comment:
In this case, I think `tx_buff` is not need but it is allocated and passed to `reqs.push()`, am I right? I think the same applies to `rx_buff` in the `else` branch.

Reply:
For read-only transfer, tx_buf is not needed, but the Linux interface requires it as null pointer, so I just allocate an empty vector for it, and this will be checked in spi.rs transfer method, if empty, then pass null pointer.